### PR TITLE
Add F# ( command dotnet-compile-fsc )

### PR DESCRIPTION
**Pull request by enricosada**
add command `dotnet-compile-fsc`

It's used with `"compilerName": "fsc"` in project.json

use`fsc.exe` from `Microsoft.FSharp.Compiler.netcore` package ( ref http://myget.org/gallery/fsharp-daily )

the FSharp.Core.netcore package is in the same feed

### CODE REVIEW

**Comment by Alxandr** [comment]
How are you dealing with file order? You should have a test project with more than 1 source file.

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json [hunk 1]
```diff
@@ -0,0 +1,31 @@
+{
+    "name": "dotnet-compile-fsc",
+    "version": "1.0.0-*",
+    "compilationOptions": {
+        "emitEntryPoint": true
+    },
+    "dependencies": {
+        "NETStandard.Library": "1.0.0-rc2-23616",
+        
+        "Microsoft.Net.Compilers.netcore": "1.2.0-beta-20151215-01",
```
Comment (agocke) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json]: These are the Roslyn compilers -- you should remove it.

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+        "Microsoft.DotNet.Cli.Utils": {
+            "type": "build",
+            "version": "1.0.0-*"
+        },
+        "Microsoft.Extensions.CommandLineUtils.Sources": {
+            "type": "build",
+            "version": "1.0.0-*"
+        },
+        "Microsoft.DotNet.Compiler.Common": "1.0.0-*"
```
Comment (agocke) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json]: I would expect a reference to a coreclr compatible F# compiler somewhere around here.

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+                commonArgs.Add("--target:library");
+            }
+
+            return commonArgs;
+        }
+
+        private static Command RunFsc(string fscArgs)
+        {
+            // Locate CoreRun
+            return Command.Create("fsc", fscArgs);
```
Comment (agocke) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs]: This doesn't seem to locate CoreRun. :)

Also, dotnet-cli needs to bring its dependencies with it. You can't rely on `fsc` being installed on the machine.

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        private static IEnumerable<string> GetDefaultOptions()
+        {
+            var args = new List<string>()
+            {
+                "--noframework",
+                "--nologo"
+            };
+
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                args.Add("--debug:full");
```
Comment (agocke) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs]: Does fsc support portable PDBs? If so, we should use that on other platforms.

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
```
Comment (agocke) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/Properties/AssemblyInfo.cs]: Is this file necessary? I don't see anything interesting happening here.

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json (hunk 2 above)
Comment (Alxandr) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json]: At present, no such thing exists afaik.

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs (hunk 4 above)
Comment (enricosada) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs]: Not portable pdb atm, in the todo list

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/Properties/AssemblyInfo.cs (hunk 5 above)
Comment (enricosada) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/Properties/AssemblyInfo.cs]: Nope, i'll remove it

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json (hunk 1 above)
Comment (enricosada) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json]: Ok

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json (hunk 2 above)
Comment (enricosada) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json]: No released version of coreclr fcs. 
We can use the desktop version installed to use dotnet cli to compile project.json. 
That can help existing projects right now to try dotnet cli

Comment (agocke) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json]: Maybe something on http://myget.org/gallery/fsharp-daily would suffice.

Otherwise, I would consider this a blocker. F# needs to be delivered with dotnet-cli and it needs to run on CoreCLR.

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs (hunk 3 above)
Comment (enricosada) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs]: I use fsc in PATH for now, no released version of fcs
At least i can use dotnet cli to compile existing f# project and try project.json
Is not the final coreclr ready implementation but is the starting point

Comment (enricosada) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs]: Btw this comment is from original csc project

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json (hunk 2 above)
Comment (enricosada) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/project.json]: added fsc coreclr from  http://myget.org/gallery/fsharp-daily

**Comment by enricosada** [comment]
ok @agocke done! 

- fsc from netcore ( fsharp daily nuget feed )
- use file ordering ( with `compileFiles` in project.json ) in test

File: src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs (hunk 3 above)
Comment (enricosada) [inline src/Microsoft.DotNet.Tools.Compiler.Fsc/Program.cs]:
```csharp
// Locate CoreRun
```

this commet was from csc implementation ( i used that as a base )

**Comment by enricosada** [comment]
about build errors:

- osx, error on git clone

```
06:22:16  > git init /Users/dotnet-bot/j/workspace/dotnet_cli/debug_osx_prtest # timeout=10
06:22:16 
ERROR: Error cloning remote repo 'origin'
```

- win release:

```
06:36:14 Compiling Microsoft.DotNet.ProjectModel for DNXCore,Version=v5.0
06:36:18 
06:36:18 Unhandled Exception: System.AccessViolationException: Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
06:36:18    at Microsoft.Cci.ISymUnmanagedWriter2.Close()
```

i get error like that sometimes on build ( on master branch too, not only in my pr ) and on different projects. dunno. if i rerun `build.cmd` more times, it build correctly

**Comment by enricosada** [comment]
btw @agocke thx for review!

**Comment by agocke** [comment]
No problem :). My plane's about to take off but I'll see if I can try it out and get this merged ASAP.

**Comment by agocke** [comment]
It looks like the FSharpTestApp and FSharpTestLibrary are not compiled in the smoke test -- we should probably add those to `smoke-test.sh`.

It also seems to fail with the following error on OS X:

```
Compiling FSharpTestApp for DNXCore,Version=v5.0

Unhandled Exception: System.ComponentModel.Win32Exception: No such file or directory
   at System.Diagnostics.Process.ResolvePath(String filename)
   at System.Diagnostics.Process.StartCore(ProcessStartInfo startInfo)
   at Microsoft.DotNet.Cli.Utils.Command.Execute()
   at Microsoft.DotNet.Tools.Compiler.Fsc.Program.Main(String[] args)

Compilation failed.
    0 Warning(s)
    0 Error(s)
```

Have you tested this on a non-Windows machine?

**Comment by enricosada** [comment]
I'll check tomorrow on my OSX but i think fcs doesnt work xplat atm. 
I'll add library to smoke test and check build log

**Comment by enricosada** [comment]
- rebased to latest `HEAD`
- added `-help` ( like `dotnet-compile-csc` )
- use `--simpleresolution`
- disable pdb on osx ( doesnt work )
- move test projects in `test/FSharpTestProjects`, same structure as `test/TestProjects`, so we can use same integration tests i think ( changing only test directory )
- add CompileFail and TestAppWithArgs test
- locate and use `corerun` to execute `fsc.exe` 
- HACK: there is no FSharp.Core nuget atm, so i use the FSharp.Core.dll bundled with fsc.exe /cc @KevinRansom
- exe doesnt work ( error error FS0208 ) /cc @KevinRansom

**Comment by KevinRansom** [comment]
The latest f# nuget package is here:

https://www.myget.org/feed/fsharp-daily/package/nuget/Microsoft.FSharp.Compiler.netcore/1.0.0-alpha-151218

**Comment by enricosada** [comment]
@agocke i need to check osx, setting up environment, i'll try tonight

**Comment by enricosada** [comment]
@KevinRansom i need a FSharp.Core nuget package, for add the FSharp.Core to library project and console project. 
i cant directly use the compiler nuget, let's speak in standup tonight :smile:

**Comment by enricosada** [comment]
ok

- **compile to exe works!!** ( TestApp, TestAppWithArgs )
- use FSharp.Core coreclr nuget ( thx @KevinRansom )
- upgrate to latest alpha of fsharp compiler
- workaround FS0208 ( ref Microsoft/visualfsharp#824 )
- use default.win32manifest ( ref Microsoft/visualfsharp#825 )

**Comment by enricosada** [comment]
ouch, all failed because

```
06:28:09 More than one runtime.json file has declared imports for 'any'
```

i see `runtimes\any\native` in `Microsoft.FSharp.Compiler.netcore.1.0.0-alpha-151218.nupkg` , but was there before, a new dependency added it?
I dont know how to fix it, locally was working before ( rebuilding from stratch atm to check )

/cc @KevinRansom @agocke @davidfowl

**Comment by KevinRansom** [comment]
Compiler now has a dependency on fsharp.core

**Comment by enricosada** [comment]
revert upgrade because `Microsoft.FSharp.Core.netcore` v1.0.0-alpha-151221 give an error with `NETStandard.Platform` v1.0.0-rc2-23616

ref Microsoft/visualfsharp#832

**Comment by piotrpMSFT** [comment]
@enricosada can you squash the commits? Then we'll go ahead and merge.

**Comment by enricosada** [comment]
@piotrpMSFT  sure, trying to build locally atm.

**Comment by enricosada** [comment]
@piotrpMSFT done, tested locally after squash

**Comment by enricosada** [comment]
```
[1m[33m/mnt/resource/j/workspace/dotnet_cli/debug_centos7.1_prtest/tools/MultiProjectValidator/tools/MultiProjectValidator/Program.cs(30,29): warning CS0168: The variable 'e' is declared but never used[39m[22m
08:10:31 
08:10:31 Unhandled Exception: System.NullReferenceException: Object reference not set to an instance of an object.
08:10:31    at ProjectSanity.AnalysisRules.DependencyMismatch.DependencyInfo.Create(ProjectContext context, ```

i dont think centos build failures are about this pr.
